phoenix-core/src/main/java/org/apache/phoenix/cache/ServerMetadataCache.java

```diff
@@ -0,0 +1,144 @@
+                    String key = notification.getKey().toString();
+                    LOGGER.debug("Expiring " + key + " because of "
+                            + notification.getCause().name());
+                })
+                // maximum number of entries this cache can handle.
+                .maximumSize(10000L)
+                .expireAfterAccess(maxTTL, TimeUnit.MILLISECONDS)
+                .build();
+    }
+
+    /**
+     * Returns the last DDL timestamp from the table. If not found in cache, then query SYSCAT regionserver.
+     * @param maintainer
+     * @return last DDL timestamp
+     * @throws Exception
+     */
+    public long getLastDDLTimestampForTable(
+            DDLTimestampMaintainersProtos.DDLTimestampMaintainer maintainer) throws IOException {
+        byte[] tenantID = maintainer.getTenantID().toByteArray();
+        byte[] schemaName = maintainer.getSchemaName().toByteArray();
+        byte[] tableName = maintainer.getTableName().toByteArray();
+        String fullTableNameStr = SchemaUtil.getTableName(schemaName, tableName);
+        byte[] tableKey = SchemaUtil.getTableKey(tenantID, schemaName, tableName);
+        ImmutableBytesPtr tableKeyPtr = new ImmutableBytesPtr(tableKey);
+        // Lookup in cache if present.
+        Long lastDDLTimestamp = lastDDLTimestampMap.getIfPresent(tableKeyPtr);
+        if (lastDDLTimestamp != null) {
+            LOGGER.trace("Retrieving last ddl timestamp value from cache for tableName: {}",
+                    fullTableNameStr);
+            return lastDDLTimestamp;
+        }
+
+        PTable table;
+        String tenantIDStr = Bytes.toString(tenantID);
+        if (tenantIDStr == null || tenantIDStr.isEmpty()) {
+            tenantIDStr = null;
+        }
+        Properties properties = new Properties();
+        if (tenantIDStr != null) {
+            properties.setProperty(TENANT_ID_ATTRIB, tenantIDStr);
+        }
+        try (Connection connection = QueryUtil.getConnectionOnServer(properties, this.conf)) {
+            // Using PhoenixRuntime#getTableNoCache since se don't want to read cached value.
+            table = PhoenixRuntime.getTableNoCache(connection, fullTableNameStr);
+            // TODO PhoenixRuntime#getTableNoCache can throw TableNotFoundException.
+            //  In that case, do we want to throw non retryable exception back to the client?
+            // Update cache with the latest DDL timestamp from SYSCAT server.
+            lastDDLTimestampMap.put(tableKeyPtr, table.getLastDDLTimestamp());
+        } catch (SQLException sqle) {
+            // Throw IOException back to the client and let the client retry depending on
+            // the configured retry policies.
+            LOGGER.warn("Exception while calling calling getTableNoCache for tenant id: {}," +
+                            " tableName: {}", tenantIDStr, fullTableNameStr, sqle);
+            throw new IOException(sqle);
```

Comment:
We are calling this from [BaseScannerRegionObserver#preScannerOpen](https://github.com/apache/phoenix/blob/2dab08504c4828119bf55ad69653477ab8085bd1/phoenix-core/src/main/java/org/apache/phoenix/coprocessor/BaseScannerRegionObserver.java#L249) and this can only throw IOException.